### PR TITLE
Support whitespace and punctuation for escaped identifiers

### PR DESCRIPTION
**Pull request by adellibovi**
Hello everyone 👋 
This is a working in progress prototype for a new evolution pitch about adding support for whitespace and punctuation for escaped identifiers.

In few words this would enable us to do something like:
```swift
func `test object returns true when validation succeed`() { ... }
```

For more details, the discussion is currently being held at: https://forums.swift.org/t/pre-pitch-whitespaces-and-punctuation-for-methods-declarations/32073

### CODE REVIEW

**Review by CodaFi — APPROVED** [review]
Approach LGTM.  I’d appreciate more regression tests but I’m not gonna block on it.  In particular
- A semantic test for (overloaded) declrefs of declarations with these names
- A syntactic test that exercises typo correction
- A syntactic test that hits the disallowed ranges in the new predicates

**Comment by CodaFi** [comment]
Let’s see what’s up

@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please build toolchain

**Review by allevato — CHANGES_REQUESTED** [review]
> more regression tests

I mentioned it in the [forum thread](https://forums.swift.org/t/whitespaces-and-punctuation-for-escaped-identifiers/32073/22?u=allevato), but one area that still needs work is `SILPrinter`. It currently only checks identifiers against the list of keywords to determine escaping, so the following:

```swift
func `foo bar baz`() {}
```

would get written out by `-emit-sil(gen)`, I believe, as

```swift
func foo bar baz()
```

and then would fail to parse if someone passed that SIL back as input to the compiler.

So it would be good to have some tests that take a file with some escaped identifiers, compile it to SIL, and then compile that, to verify the functionality that still needs to be implemented.

**Comment by adellibovi** [comment]
Thanks @CodaFi for the review, I had to push few more commits, if you feel like that requires another review of the PR I will be happy to receive your feedbacks 😊 Can I kindly ask you to execute the tests and the tool chain again? Thanks!

File: lib/Parse/Lexer.cpp
Comment (rintaro) [inline lib/Parse/Lexer.cpp]: Did you mean `'\n'`?

Comment (adellibovi) [reply lib/Parse/Lexer.cpp]: oops, yes, pushed the fix :)

**Comment by jckarter** [comment]
@swift-ci Please build toolchain

**Comment by nonsensery** [comment]
Not sure if this has been fixed in the mean time, but I think I encountered a bug in toolchain dc087f5.

When I write this code:

```swift
func `some weird method`(`argument one`: Int, `argument two`: Int) -> Int {
    `argument one` + `argument two`
}

`some weird method`(`argument one`: 1, `argument two`: 2)
```

I get warnings that the arguments don't need to be escaped, If I apply the fixit, it removes the backticks and the code won't compile.

Screen Shot 2020-01-24 at 2 03 03 PM

**Comment by adellibovi** [comment]
Closing as the proposal got rejected, thanks for the PR review anyway!

**Pull request closed without merging**
